File: assignments/java/src/main/java/membermanagement/MemberManagement.java
package membermanagement;

import java.util.*;

public class MemberManagement {
    private Scanner sc = new Scanner(System.in);

    private int totalCnt = 0;
    private int memberCnt = 0;
    private Member[] members;

    private int printPricePlan() {
        System.out.println("================== 요금제 선택 ==================");
        System.out.println("[1]Lite : 10명 [2]Basic : 20명 [3]Premium : 30명");
        System.out.println("===============================================");
        System.out.print("> ");

        return Integer.parseInt(sc.nextLine());
    }

    private int printMenu() {
        System.out.println("===============================================");
        System.out.println("수행할 업무 선택 - 현재 회원수: " + memberCnt +"/" + totalCnt);
        System.out.println("[1]회원추가 [2]회원조회(메일) [3]회원조회(이름)");
        System.out.println("[4]회원전체조회 [5]회원정보 수정 [6]회원삭제");
        System.out.println("[7]프로그램 종료");
        System.out.println("===============================================");
        System.out.print("> ");

        return Integer.parseInt(sc.nextLine());
    }

    private void addMember() {
        if (memberCnt == members.length) {
            System.out.println("------------------- 정원 초과 -------------------");
            return;
        }

        System.out.println("-----------------------------------------------");
        System.out.print("이름 > ");
        String name = sc.nextLine();
        System.out.println("-----------------------------------------------");
        System.out.print("이메일 > ");
        String email = sc.nextLine();

        if (checkEmail(email)) {
            System.out.println("이미 존재하는 이메일");
            return;
        }

        System.out.println("-----------------------------------------------");
        System.out.print("전화번호 > ");
        String phone = sc.nextLine();

        members[memberCnt] = new Member(name, email, phone);
        memberCnt++;
    }

    private boolean checkEmail(String email) {
        for (int i=0; i<memberCnt; i++) {
            if (email.equals(members[i].getEmail())) {
                return true;
            }
        }
        return false;
    }

    private void selectEmail() {
        System.out.println("-----------------------------------------------");
        System.out.print("이메일 > ");
        String email = sc.nextLine();

        for (int i=0; i<memberCnt; i++) {
            if (email.equals(members[i].getEmail())) {
                System.out.println("이름: " + members[i].getName() + " 이메일: " + members[i].getEmail() + " 전화번호: " + members[i].getPhone());
                return;
            }
        }

        System.out.println("존재하지 않는 이메일");
    }

    private void selectName() {
        System.out.println("-----------------------------------------------");
        System.out.print("이름 > ");
        String name = sc.nextLine();

        boolean found = false;

        for (int i=0; i<memberCnt; i++) {
            if (name.equals(members[i].getName())) {
                System.out.println("이름: " + members[i].getName() + " 이메일: " + members[i].getEmail() + " 전화번호: " + members[i].getPhone());
                found = true;
            }
        }

        if (!found) {
            System.out.println("존재하지 않는 이름");
        }
    }

    private void selectAll() {
        for (int i=0; i<memberCnt; i++) {
            System.out.println("이름: " + members[i].getName() + " 이메일: " + members[i].getEmail() + " 전화번호: " + members[i].getPhone());
        }
    }

    private void updateMember() {
        System.out.println("-----------------------------------------------");
        System.out.print("이메일 > ");
        String email = sc.nextLine();

        int idx = -1;

        for (int i=0; i<memberCnt; i++) {
            if (email.equals(members[i].getEmail())) {
                idx = i;
                break;
            }
        }

        if (idx == -1) {
            System.out.println("존재하지 않는 회원");
            return;
        }

        System.out.println("-----------------------------------------------");
        System.out.print("이름 > ");
        members[idx].setName(sc.nextLine());
        System.out.println("-----------------------------------------------");
        System.out.print("이메일 > ");
        members[idx].setEmail(sc.nextLine());
        System.out.println("-----------------------------------------------");
        System.out.print("전화번호 > ");
        members[idx].setPhone(sc.nextLine());
    }

    private void deleteMember() {
        System.out.println("-----------------------------------------------");
        System.out.print("이메일 > ");
        String email = sc.nextLine();

        int idx = -1;

        for (int i=0; i<memberCnt; i++) {
            if (email.equals(members[i].getEmail())) {
                idx = i;
                break;
            }
        }

        if (idx == -1) {
            System.out.println("존재하지 않는 회원");
            return;
        }

        for (int i=idx; i<memberCnt-1; i++) {
            members[i] = members[i + 1];
        }

        memberCnt--;
        members[memberCnt] = null;
    }

    public void run() {
        int num = printPricePlan();

        while (num < 1 || num > 3) {
            System.out.println("올바른 요금제 선택!.");
            num = printPricePlan();
        }

        members = new Member[num*10];
        totalCnt = num * 10;

        while (true) {
            int choice = printMenu();

            switch (choice) {
                case 1:
                    addMember();
                    break;
                case 2:
                    selectEmail();
                    break;
                case 3:
                    selectName();
                    break;
                case 4:
                    selectAll();
                    break;
                case 5:
                    updateMember();
                    break;
                case 6:
                    deleteMember();
                    break;
                case 7:
                    System.out.println("================= 프로그램 종료 =================");
                    return;
                default:
                    System.out.println("=============== 올바른 번호 입력! ===============");
            }
        }
    }
}

File: assignments/kotlin/src/main/kotlin/membermanagement/MemberManagement.kt
package membermanagement

var totalCnt = 0
var memberCnt = 0

fun printPricePlan(): Int {
    println("===============================================")
    println("[요금제 선택]")
    println("[1]Lite : 10명 [2]Basic : 20명 [3]Premium : 30명")
    print("> ")

    return readln().toInt()
}

fun printMenu(): Int {
    println("===============================================")
    println("[수행할 업무 선택 - 현재 회원 수: $memberCnt/$totalCnt]")
    println("[1]회원추가 [2]회원조회(메일) [3]회원조회(이름)")
    println("[4]회원전체조회 [5]회원정보 수정 [6]회원삭제")
    println("[7]프로그램 종료")
    print("> ")

    return readln().toInt()
}

fun addMember(members: Array<Array<String>>) {
    if (memberCnt == totalCnt) {
        println("-----------------------------------------------")
        println("회원 초과!")

        return
    }

    println("-----------------------------------------------")
    print("이름 입력: ")
    val name = readln()
    print("이메일 입력: ")
    val email = readln()
    print("연락처 입력: ")
    val phone = readln()

    if (findIndex(members, 1, email) != -1) {
        println("-----------------------------------------------")
        println("이미 존재하는 회원")

        return
    }

    members[memberCnt][0] = name
    members[memberCnt][1] = email
    members[memberCnt][2] = phone
    memberCnt++
    println("-----------------------------------------------")
    println("회원 등록 완료")
}

fun printMember(member: Array<String>) {
    println("[이름] ${member[0]}, [이메일] ${member[1]}, [연락처] ${member[2]}")
}

fun findIndex(members: Array<Array<String>>, col: Int, value: String): Int {
    for (i in 0 until memberCnt) {
        if (value == members[i][col]) {
            return i
        }
    }

    return -1
}

fun selectEmail(members: Array<Array<String>>) {
    println("-----------------------------------------------")
    print("이메일 입력: ")
    val email = readln()

    val idx = findIndex(members, 1, email)

    if (idx == -1) {
        println("-----------------------------------------------")
        println("정보 없음!")

        return
    }

    printMember(members[idx])
}

fun selectName(members: Array<Array<String>>) {
    println("-----------------------------------------------")
    print("이름 입력: ")
    val name = readln()

    val idx = findIndex(members, 0, name)

    if (idx == -1) {
        println("-----------------------------------------------")
        println("정보 없음!")

        return
    }

    printMember(members[idx])
}

fun selectAll(members: Array<Array<String>>) {
    if (memberCnt == 0) {
        println("-----------------------------------------------")
        println("회원 없음")

        return
    }

    for (i in 0 until memberCnt) {
        print("${i + 1}. ")
        printMember(members[i])
    }
}

fun updateMember(members: Array<Array<String>>) {
    println("-----------------------------------------------")
    print("수정할 회원의 이메일 입력: ")
    val email = readln()

    val idx = findIndex(members, 1, email)

    if (idx == -1) {
        println("-----------------------------------------------")
        println("회원 없음")

        return
    }

    println("-----------------------------------------------")
    print("현재 정보 → ")
    printMember(members[idx])

    println("-----------------------------------------------")
    print("새 이름 입력: ")
    members[idx][0] = readln()
    print("새 이메일 입력: ")
    members[idx][1] = readln()
    print("새 연락처 입력: ")
    members[idx][2] = readln()

    println("-----------------------------------------------")
    println("수정 완료")
}

fun deleteMember(members: Array<Array<String>>) {
    println("-----------------------------------------------")
    print("삭제할 회원 이메일 입력: ")
    val email = readln()

    val idx = findIndex(members, 1, email)

    if (idx == -1) {
        println("-----------------------------------------------")
        println("회원 없음")

        return
    }

    for (i in idx until memberCnt-1) {
        members[i][0] = members[i+1][0]
        members[i][1] = members[i+1][1]
        members[i][2] = members[i+1][2]
    }

    memberCnt--

    members[memberCnt][0] = ""
    members[memberCnt][1] = ""
    members[memberCnt][2] = ""

    println("-----------------------------------------------")
    println("삭제 완료")
}

fun main() {
    val num = printPricePlan()
    totalCnt = num * 10
    val members = Array(totalCnt) { Array(3) { "" } }

    while (true) {
        when (printMenu()) {
            1 -> addMember(members)
            2 -> selectEmail(members)
            3 -> selectName(members)
            4 -> selectAll(members)
            5 -> updateMember(members)
            6 -> deleteMember(members)
            7 -> {
                println("-----------------------------------------------")
                println("프로그램 종료")
                println("===============================================")

                return
            }
            else -> {
                println("-----------------------------------------------")
                println("올바른 번호 입력!")
            }
        }
    }
}